File: perf/benchmarks/micro/deque_ops/deque_ops.go
// deque_ops — dynamic-array push/pop micro-benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func main() {
	n := int64(2_000_000)
	if len(os.Args) > 1 {
		v, _ := strconv.ParseInt(os.Args[1], 10, 64)
		n = v
	}
	t0 := time.Now()
	a := make([]int64, 0)
	var seed int64 = 1
	for i := int64(0); i < n; i++ {
		seed = (seed*1103515245 + 12345) & 0x7FFFFFFF
		a = append(a, seed)
	}
	var cs int64 = 0
	for len(a) > 0 {
		cs ^= a[len(a)-1]
		a = a[:len(a)-1]
	}
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/micro/deque_ops/deque_ops.rs
// deque_ops — dynamic-array push/pop micro-benchmark (Rust).
use std::env;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: i64 = if args.len() > 1 { args[1].parse().unwrap_or(2_000_000) } else { 2_000_000 };
    let t0 = Instant::now();
    let mut a: Vec<i64> = Vec::new();
    let mut seed: i64 = 1;
    for _ in 0..n {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7FFF_FFFF;
        a.push(seed);
    }
    let mut cs: i64 = 0;
    while let Some(v) = a.pop() {
        cs ^= v;
    }
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
